### PR TITLE
Last year of updates on ItMightBeKaraoke branch

### PR DESCRIPTION
**Pull request by ItMightBeKaraoke**
Sorry it's been so long, but I think this is finally mergeable. There are a few differences in the default behavior from your original version:

- Gaps are added in if present in the sync (e.g. if there is a 50ms pause between two syllables, it will add a {\k5} before the \kf or \k for the next syllable). With the current version, it will just eat the gap and the line will end too soon.
- Font is scaled up a bit (1.4x) to roughly accommodate the difference between font size (KBS) and line height (ASS). I do want to improve this, at least offering it as a parameter, but haven't gotten to it yet
- Some escaping is added to lyrics from .kbp:
  - "{~}" to "/" - this is how KBS writes "/", because it's a special character in its file format
  - "{}" prepended with "\\" - This is not technically part of the .ass standard, but is supported by libass. It's either we do this, just delete {}. or allow arbitrary .ass code, so this seemed like the best option to me. Maybe it should be made configurable too
  - Add a zero-width space between \ and something that would generate an escape sequence to display it literally
  - Space at the beginning of the line changed to "\h"
- If the user has an offset configured in KBS, it will apply it during the conversion (configurable to disable that or use a different offset)
- A couple bugfixes related to styles - invalid style numbers default to style 0, and gaps in styles (e.g. first two styles are present, third is missing, then fourth is present) do not cause problems

Of course, there are also a ton of additional options used by kbp2video users to create YouTube videos. In CLI usage, --full-mode (-f) enables several of these at once.

### CODE REVIEW

File: src/types.d.ts [hunk 1]
```diff
@@ -19,7 +30,82 @@ interface StyleElement {
...
+  interface IDialogueKV<T> {
+	key: T;
+	value: IDialogue;
+  }
+  interface IFormatKV {
+	key: "Format";
+	value: Array<keyof IDialogue>;
+  }
+  
+  // ? same as sentence ?
```
Comment (Aeden-B) [inline src/types.d.ts]: I don't understand this comment ?

File: package.json
Comment (Aeden-B) [inline package.json:1]: You can add a
  "contributors": [
    "Matt M",
  ],
  
  or   "contributors": [
    "Matt M <your email>",
  ],

after the author field

File: src/index.ts [hunk 2]
```diff
@@ -1,127 +1,268 @@
...
-	}
-	const txtFile = process.argv[2];
+  }
+
+	// Per yargs docs, use of terminalWidth() in typescript requires workaround
+	// of assigning a variable name to the instance so it can be referenced
+	const yargsInstance = yargs(hideBin(process.argv))
+
+	const argv = yargsInstance
+		.scriptName('kbp2ass')
```
Comment (Aeden-B) [inline src/index.ts]: Quotes are not very consistent, do you want to use double or single ?
Same for indent do you want to have 2 or 4 space or use tabs ?

File: src/kbp.ts [hunk 3]
```diff
@@ -158,15 +227,28 @@ export default class KBPParser {
 				syllable.text = matches[0];
 
 				// Add the start time of the syllable
-				syllable.start = Math.floor(parseInt(matches[1].trim()) * 10);
+
+				// TODO: Better handling of fixed text. Currently this will just set
```
Comment (Aeden-B) [inline src/kbp.ts:230]: How do you know in KBP when a text is fixed ?

**Review by Aeden-B — CHANGES_REQUESTED** [review]

File: src/types.d.ts (hunk 1 above)
Comment (ItMightBeKaraoke) [reply src/types.d.ts]: Good question. It was part of Ian's big type-ification commit and I must not have noticed it. It doesn't mean anything to me so I'll remove it.

File: src/index.ts (hunk 2 above)
Comment (ItMightBeKaraoke) [reply src/index.ts]: Good point. I think your original code was tabs, so I can standardize on that throughout. I think it's mostly there except for the types file which ended up with a strange 2 space indentation, with tabs instead of 4 spaces. There were probably a few places where my normal editor config of 4-space indents leaked in that I can fix as well.

As far as quotes, I can fix that too. Do you have a preference?

File: src/kbp.ts (hunk 3 above)
Comment (ItMightBeKaraoke) [reply src/kbp.ts:230]: It's in the style field. Uppercase is wiped lines, and lowercase is fixed. For example, this line would have style 00 with wiping:

`C/A/3648/4349/0/15/0`

This line would be the same but fixed:

`C/a/3648/4349/0/15/0`

The issue is that in .kbp, a fixed line uses the pre-wipe colors whereas in .ass a line without wiping uses the PrimaryColour, so we need to do one of the following:

1. Create an additional style for fixed that sets the primary color to the pre-wipe color and optionally also remove the wiping
2. Add an override tag for the color on each fixed line (and optionally also remove the wiping)
3. Set the wiping to occur after the line is gone (solution I chose)

The advantage to (3) is that if you already have timings on the line and decide you do in fact want to wipe it, you can just change the delay in the initial \k tag. But it's otherwise kind of hacky.

File: src/index.ts (hunk 2 above)
Comment (ItMightBeKaraoke) [reply src/index.ts]: Indent should be fixed, all tabs now. Not sure which way to go with quotes since it's about 50/50 now

Comment (ItMightBeKaraoke) [reply src/index.ts]: I went ahead with double

Comment (Aeden-B) [reply src/index.ts]: no problem for me but can you update to "double" the eslint config https://github.com/ItMightBeKaraoke/kbp2ass/blob/main/.eslintrc.json#L26 ?

File: src/kbp.ts (hunk 3 above)
Comment (Aeden-B) [reply src/kbp.ts:230]: You can also surcharge the style color for a line with \c or \2c tag if it's useful https://aegisub.org/docs/latest/ass_tags/, see "Set color"

File: src/index.ts (hunk 2 above)
Comment (ItMightBeKaraoke) [reply src/index.ts]: This should be properly fixed now:

  $ npx eslint src/*.ts
  $

Note that I added avoidEscape: true as well so it would allow e.g. 'foo "bar" baz' instead of requiring "foo \\"bar\\" baz".

File: src/kbp.ts (hunk 3 above)
Comment (ItMightBeKaraoke) [reply src/kbp.ts:230]: Right, that's what I meant by (2). Do you have any preference between (1) or (2)? (2) would definitely be easier to implement with the way the code is currently structured.

Also, if PrimaryColour and SecondaryColour are the same, visually it doesn't matter if the wipe tags are on the line or not. It would look a little cleaner if we removed them, but leaving them in place would allow wiping to be turned back on if desired by just removing the \\1c tag.

Comment (Aeden-B) [reply src/kbp.ts:230]: In our case, we prefer to have specific styles if possible.
But you are the main user of this kbp2ass converter so feel free to choose what you prefer.

**Review by Aeden-B — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -26,26 +26,106 @@ Options are :
 
 ```JS
 {
-  syllable_precision: boolean,
-  minimum_progression_duration: number
-
+  'syllable-precision': boolean,
```
Comment (Aeden-B) [inline README.md:29]: @ItMightBeKaraoke Do you know why this change ? If it's the _ character the problem, I would prefer to avoid the - character. Can we use syllablePrecision instead ?

Comment (Aeden-B) [reply README.md:29]: it's also not consistent with dialogueScript
